### PR TITLE
WIP: Adds `time` to schema & refactors Gemini loaders

### PR DESCRIPTION
**Pull request by ashfurrow**
Follow [a discussion on Slack](https://artsy.slack.com/archives/C0C4AJ1PF/p1510775689000184), this PR started as solely the addition of `time` to the GraphQL schema. However, after exploring the idioms a bit, I thought it might make sense to refactor the Gemini loaders to make the `loaderOneOffFactory` have the same shape as the other loader factories. If that's not something we want to pursue, we can just drop https://github.com/artsy/metaphysics/commit/3fa755d611711a21957dce4321ad8ed05e08ebe2 from the PR.

screen shot 2017-11-15 at 9 47 38 pm

WIP because I still need to test that the Gemini loaders actually work 🙊

### CODE REVIEW

File: schema/asset_uploads/create_asset_request_mutation.js [hunk 1]
```diff
@@ -93,6 +93,6 @@ export default mutationWithClientMutationId({
   },
   mutateAndGetPayload: ({ name, acl }, request, { rootValue: { createNewGeminiAssetLoader } }) => {
     if (!createNewGeminiAssetLoader) return null
-    return createNewGeminiAssetLoader({ name, acl })
+    return createNewGeminiAssetLoader(name, acl)
```
Comment (damassi) [inline schema/asset_uploads/create_asset_request_mutation.js]: IMO named parameters are always better than the other way around as there are no ordering restrictions and intellisense is better. Why did you remove?

File: schema/asset_uploads/finalize_asset_mutation.js [hunk 2]
```diff
@@ -50,6 +50,6 @@ export default mutationWithClientMutationId({
     { rootValue: { createNewGeminiEntryAssetLoader } }
   ) => {
     if (!createNewGeminiEntryAssetLoader) return null
-    return createNewGeminiEntryAssetLoader({ name, template_key, source_key, source_bucket, metadata })
+    return createNewGeminiEntryAssetLoader(template_key, source_key, source_bucket, metadata)
```
Comment (damassi) [inline schema/asset_uploads/finalize_asset_mutation.js]: Same here -- this is less flexible than before.

File: schema/time/index.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    zone: { type: GraphQLString },
+    iso8601: { type: GraphQLString },
+  }
+}
+
+export const TimeType = new GraphQLObjectType({
+  name: "Time",
+  fields: () => ({
+    ...timeFields(),
+  }),
```
Comment (damassi) [inline schema/time/index.js]: Could shorten this to `fields: () => timeFields()` unless additional properties expect to be merged.

File: schema/asset_uploads/create_asset_request_mutation.js (hunk 1 above)
Comment (ashfurrow) [reply schema/asset_uploads/create_asset_request_mutation.js]: Hmm, that's true. I removed them because they are required, but if the convention is to prefer objects, I'll conform to that 👍

File: schema/time/index.js (hunk 3 above)
Comment (ashfurrow) [reply schema/time/index.js]: Ah yeah, this was just left over from another part of the codebase I was referencing (which added the `cached` field).

File: schema/asset_uploads/create_asset_request_mutation.js (hunk 1 above)
Comment (damassi) [reply schema/asset_uploads/create_asset_request_mutation.js]: I don't think its a convention but with ES6's object shorthand syntax its much more common these days

File: schema/time/index.js (hunk 3 above)
Comment (alloy) [reply schema/time/index.js]: Or just even `fields: timeFields`, seeing as it is a function that returns an object of fields.

**Comment by alloy** [comment]
I’m slightly hesitant to add a `time` root field, as it’s not really a first-class domain model. How about exposing the field on the types that would usually need it? I don’t know how the data is being used to render a UI, but I could imagine e.g.

```graphql
{
  sale(id: "sothebys-latin-america-modern-art") {
    system_time
  }
}
```

Or if there are many more system synchronisation thingies besides time that I’m not aware of (?), then _maybe_ we can model those under a `system` root field.

File: lib/loaders/api/loader_one_off_factory.js [hunk 4]
```diff
@@ -7,12 +7,19 @@ import { loaderInterface } from "./loader_interface"
  * This factory provides a short-cut system for our data loader system, it provides
  * and uncached result that is close to the equivilent to calling the fetch request directly.
```
Comment (alloy) [inline lib/loaders/api/loader_one_off_factory.js]: Reading this doesn’t really tell me what it does nor do I like the name for the same reason. Is it the same as an unauthenticated data loader, in that it does cache per query execution but _not_ to memcache? Or is it completely different in other ways? /cc @orta

(PS There’s some spelling errors here.)

File: lib/loaders/loaders_without_authentication/gravity.js [hunk 5]
```diff
@@ -2,9 +2,10 @@
 import factories from "../api"
 
 export default requestID => {
-  const { gravityLoaderWithoutAuthenticationFactory } = factories(requestID)
-  const gravityLoader = gravityLoaderWithoutAuthenticationFactory
-
+  const {
+    gravityLoaderWithoutAuthenticationFactory: gravityLoader,
+    gravityLoaderWithoutAuthenticationWithoutCachingFactory: oneOffGravityLoader,
```
Comment (alloy) [inline lib/loaders/loaders_without_authentication/gravity.js]: Here’s a good example of where the ambiguity and naming leads us astray. Data loaders _should_ usually ‘cache’, just maybe only for the duration of the query execution and **not** to memcache.

File: schema/time/index.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+})
+
+const Time = {
+  type: TimeType,
+  description: "Artsy system time, never cached",
```
Comment (alloy) [inline schema/time/index.js]: I don’t think you need to include a detail like that it’s never cached. I would, however, like the description to tell the reader what to use it for, i.e. clock synchronisation.

File: schema/time/index.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    dst: { type: GraphQLBoolean },
+    unix: { type: GraphQLInt },
+    utc_offset: { type: GraphQLInt },
+    zone: { type: GraphQLString },
+    iso8601: { type: GraphQLString },
+  }
+}
+
+export const TimeType = new GraphQLObjectType({
+  name: "Time",
```
Comment (alloy) [inline schema/time/index.js]: I think the name is too broad and ambiguous. I would expect a ‘Time’ type to be able to represent any sort of time, but this type (the resolver at least) is specifically about Artsy’s system time.

Comment (ashfurrow) [reply schema/time/index.js]: Good point! What do you think about embedding it with `system`? So the query would look like:

```
{
  system {
    time { ... }
  }
}
```

(In addition t renaming `TimeType` to `SystemTimeType`?)

File: lib/loaders/api/loader_one_off_factory.js (hunk 4 above)
Comment (orta) [reply lib/loaders/api/loader_one_off_factory.js]: I built it to basically to conform to the same interfaces as the auth/unauth factories in user-code, but to skip any caching anywhere

File: lib/loaders/loaders_without_authentication/gravity.js (hunk 5 above)
Comment (ashfurrow) [reply lib/loaders/loaders_without_authentication/gravity.js]: Hmm, I'm not sure I follow. Can you expand on this? Or are there some docs/blogs I should familiarize myself with?

File: schema/time/index.js (hunk 7 above)
Comment (orta) [reply schema/time/index.js]: This makes sense to me

**Comment by ashfurrow** [comment]
@alloy I'm not opposed to moving this to the `sale` field but I worry about having to expose identical system times in other parts of the schema. To me, the "system time" concept  belongs to the Artsy API itself and I wouldn't want to limit its scope to places where we already use it. That is to say, I'd like to make accessing the system time as easy as possible to avoid developers relying on client time in any way.

I've looked through Eigen and have found uses of the system time in the following places:

- Determining if a sale is upcoming, open, live open, or closed.
- Calculating remaining time before a sale opens, live opens, or closes.
- Determining if an auth token has expired or not.
- Determining if site hero units are active.

But I don't feel strongly about this, let me know what you think 👍

**Comment by alloy** [comment]
@ashfurrow I totally agree with you that it should be easy to use in all the places where it makes sense, but when designing a graph that exposes our dataset I think we should be deliberate in exposing it only in the places where it used, even if that means having to add the field to multiple places. A ‘time’ is not really a thing in the context of Artsy’s data.

File: lib/loaders/api/loader_one_off_factory.js (hunk 4 above)
Comment (ashfurrow) [reply lib/loaders/api/loader_one_off_factory.js]: Maybe the data loaders shape changed since then? As I mentioned in the PR description, if we're adverse to the refactor we can just drop this commit: https://github.com/artsy/metaphysics/commit/3fa755d611711a21957dce4321ad8ed05e08ebe2

**Comment by ashfurrow** [comment]
Okay, sounds good. I see your point about caching for the duration of the request too.

File: lib/loaders/loaders_without_authentication/gravity.js (hunk 5 above)
Comment (alloy) [reply lib/loaders/loaders_without_authentication/gravity.js]: For the duration of the query being executed, you don’t ever want to make more than 1 request to the Gravity endpoint, right? So its value should be cached for that short period in time, but not stored in memcache and thus shared with other queries being executed.

Comment (alloy) [reply lib/loaders/loaders_without_authentication/gravity.js]: Maybe this helps https://github.com/facebook/dataloader#creating-a-new-dataloader-per-request

**Comment by alloy** [comment]
Ace 👌

File: lib/loaders/loaders_without_authentication/gravity.js (hunk 5 above)
Comment (ashfurrow) [reply lib/loaders/loaders_without_authentication/gravity.js]: Cool, that makes a lot of sense, thanks!

**Comment by ashfurrow** [comment]
I made this PR in anticipation of a Force refactor (a major one, if I'm not mistaken) instead of as a requirement for something we need to ship imminently. After discussing the idea, we now have a good sense for how we want this data exposed in the schema as well as a strategy for using data loaders in this case. I'm going to close this PR and we can revisit the conversation here when there is engineering availability for the Force change and adding `time` to the schema is required. Thank you all for the fruitful discussion, I learned a lot and I know it will be helpful going forward.
